x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlConfigMigrator.java

```diff
@@ -146,11 +146,30 @@ public void migrateConfigsWithoutTasks(ClusterState clusterState, ActionListener
                 }
         );
 
+        if (firstTime.get()) {
+            snapshotMlMeta(MlMetadata.getMlMetadata(clusterState), ActionListener.wrap(
+                    response -> {
+                        firstTime.set(false);
+                        unMarkMigrationInProgress.onResponse(Boolean.TRUE);
+                    },
+                    unMarkMigrationInProgress::onFailure
+            ));
+            return;
+        }
+
+        Version minNodeVersion = clusterState.nodes().getMinNodeVersion();
```

Comment:
I think this check should come before the backup, because then the backup will be done as close as possible before the upgrade.

Reply:
This change does not continue to start the migration after writing the snapshot instead it has to wait for the next clusterstate change event. In the upgrade tests I wasn't getting enough state changes to write the snapshot and start migration. The change needed is to call trigger migration after the write.

Reply:
I made a change to call the migration process after writing the snapshot which meant I could move the check as you suggested. Hopefully the new change will help with failing tests